### PR TITLE
Do not force refresh when write indexing buffer

### PR DESCRIPTION
**Pull request by dnhatn**
Today we periodically check the indexing buffer memory every 5 seconds or after we have used 1/30 of the configured memory. If the total used memory is over the threshold, then we refresh the "largest" shards. If refreshing takes longer these intervals (i.e., 5s or 1/30 buffer), then we continue to enqueue refreshes to these shards. This leads to two issues:

- The refresh thread pool can be exhausted and other shards can't refresh
- Execute too many refreshes for the "largest" shards

With this change, we only refresh the largest shards if they are not refreshing. Here we rely on the periodic check to trigger another refresh if needed. We can harden this by making the ongoing refresh triggers the memory check when it's completed. I opted out this option in this PR for simplicity.

See: https://discuss.elastic.co/t/write-queue-continue-to-rise/213652/

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerTests.java [hunk 1]
```diff
@@ -346,117 +328,83 @@ public void testThrottling() throws Exception {
...
+            config.getSimilarity(), new CodecService(null, logger), config.getEventListener(), config.getQueryCache(),
+            config.getQueryCachingPolicy(), config.getTranslogConfig(), config.getFlushMergesAfter(),
+            config.getExternalRefreshListener(), internalRefreshListener, config.getIndexSort(),
+            config.getCircuitBreakerService(), config.getGlobalCheckpointSupplier(), config.retentionLeasesSupplier(),
+            config.getPrimaryTermSupplier(), config.getTombstoneDocSupplier());
+    }
 
-        // TODO: would be cleaner if I could pass this 1kb setting to the single node this test created....
-        IndexingMemoryController imc = new IndexingMemoryController(settings, null, null) {
+    public void testSkipRefreshIfShardIsRefreshingAlready() throws Exception {
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerTests.java]: I only added this new test. Other tests are unchanged.

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerIT.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        };
+
+        for (int i = 0; i < 100; i++) {
+            String id = Integer.toString(i);
+            client().prepareDelete("index", id).get();
+        }
+
+        final long indexingBufferBytes1 = shard.getIndexBufferRAMBytesUsed();
+
+        imc.forceCheck();
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerIT.java]: I think there is a (very small) risk that the `IndexingMemoryController` created by the node was triggered on one  of the deletes and that this causes the writeIndexingBuffer/refresh call triggered here to return immediately. It does take a lot of bad circumstances though (of which the heap size alone will likely prevent this), but I still think we should address it. Maybe the TODO above is doable if this test is moved to its own class? Or we could simply add a node-setting that essentially disables it by setting the limit very high?

**Review by henningandersen — APPROVED** [review]
LGTM.

I left a few smaller comments. Also, it would be good to have a few extra test runs of the server module to ensure it does not introduce spurious test failures.

File: server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerTests.java [hunk 3]
```diff
@@ -346,117 +328,81 @@ public void testThrottling() throws Exception {
...
-
-        imc.forceCheck();
-
-        // We must assertBusy because the writeIndexingBufferAsync is done in background (REFRESH) thread pool:
         assertBusy(() -> {
-            try (Engine.Searcher s2 = shard.acquireSearcher("index")) {
-                // 100 buffered deletes will easily exceed our 1 KB indexing buffer so it should trigger a write:
-                final long indexingBufferBytes2 = shard.getIndexBufferRAMBytesUsed();
-                assertTrue(indexingBufferBytes2 < indexingBufferBytes1);
+            for (ThreadPoolStats.Stats stats : threadPool.stats()) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerTests.java]: ++. Adjusted in https://github.com/elastic/elasticsearch/pull/50769/commits/747551513e5712883168a48d4916d2ec6e4202e7.

File: server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerIT.java (hunk 2 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerIT.java]: Good point. I pushed https://github.com/elastic/elasticsearch/pull/50769/commits/4d04e97c214626fd184de151d6fe36949e0267cb.

**Comment by dnhatn** [comment]
> Also, it would be good to have a few extra test runs of the server module to ensure it does not introduce spurious test failures.

Yep, I am running it on my CI now.

**Comment by dnhatn** [comment]
@ywelsch @henningandersen Thanks for reviewing.

**Pull request merged**